### PR TITLE
Add multi-cell tripleo setup for adoption devel

### PR DESCRIPTION
**Pull request by bogdando**
In order to keep the HW requirments for development
of multi-cell OSP 17.1 adoption for RHOSO 18, provide
a reduced multi-stack footprint
(which is supported in tripleo, yet not in OSP):

undercloud: 1 VM
overcloud: Controller0 ( 1 VM, no HA)
cell1: Compute0, Compute1 (CellController1) ( 2 VMs)
cell2: Compute2+CellController2 (AIO VM host)

No Ceph/HCI support yet, TBD.
Only a fixed number of cells (2 extra cells) supported yet.

Add partial local dev envs support (given VMs preprovisioned):
* Fix HOME paths to match the default user home from ssh opts
  (when it is 'zuul', the commands fail on the control node,
   when executed from another user name)
* Add missing repo setup commands and notes from standalone.sh
* Support RH registry auth env vars
* Disable undercloud install validations (to allow deployments
  on low storage envs)
* Fix tripleo network config to let it configure nodes in a local
  setup, where there is no pre-configured networks
* Unplug External network from roles file as CI infra does not
  provide a gateway for it, so tripleo networking fails ping test

Add EDPM_CONFIGURE_NETWORKING to control tripleo networking
configuration. Disable it for CI jobs. Should be enabled for
local deployments.

Add j2 bool filter support to jinja_render common func.

Required-By: https://review.rdoproject.org/r/c/rdo-jobs/+/53192

JIRA [OSPRH-6548](https://issues.redhat.com/browse/OSPRH-6548)

### CODE REVIEW

File: devsetup/tripleo/overcloud_services _cell.j2
Comment (marios) [inline devsetup/tripleo/overcloud_services _cell.j2:1]: whitespace in the file name

**Review by marios — COMMENTED** [review]
please lets discuss this during tomorrow data plane call before proceeding down this path

i want us all to agree on the direction.

i think you are proposing that we will have a 3 compute hci ceph job, and then a 1 cntrol/2 compute for cells (i.e. this job) but we have not had that conversation yet. 

we need to be on the same page about the ci direction since we (ci team) are currently working here

cc @cescgina @frenzyfriday @jistr

File: devsetup/tripleo/config-download-multistack.yaml
Comment (bogdando) [inline devsetup/tripleo/config-download-multistack.yaml:146]: those need to be configurable, as in CI we use ``standalone_gateway: 192.168.122.10``

**Comment by bogdando** [comment]
got the central overcloud stack deployed locally with this, some adjustments will be needed to have the gateway_ip paramterized

File: devsetup/tripleo/config-download-multistack.yaml
Comment (bogdando) [inline devsetup/tripleo/config-download-multistack.yaml:145]: TODO: have this one j2 templated as well

File: devsetup/scripts/tripleo.sh
Comment (marios) [inline devsetup/scripts/tripleo.sh:70]: if we run this in ci we may leak the user and pass for the registry...

File: devsetup/tripleo/net_config.j2
Comment (marios) [inline devsetup/tripleo/net_config.j2:29]: as previously discussed can you please avoid making changes here that affect the 3control/3compute multinode. This file is used for the overcloud node config in that job.

File: devsetup/tripleo/tripleo_install.sh
Comment (marios) [inline devsetup/tripleo/tripleo_install.sh:33]: why these filenames out0.yaml please restore?

File: devsetup/tripleo/tripleo_install.sh
Comment (marios) [inline devsetup/tripleo/tripleo_install.sh:172]: needs newline

**Review by marios — CHANGES_REQUESTED** [review]

File: devsetup/tripleo/net_config.j2
Comment (marios) [reply devsetup/tripleo/net_config.j2:29]: (actually this is used for the undercloud node, not overcloud, but same point applies

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:70]: this is intented for local setups, but I will add ``set +x``

File: devsetup/tripleo/net_config.j2
Comment (bogdando) [reply devsetup/tripleo/net_config.j2:32]: Could we do this for multi-node as well? This change is not breaking, while allows not duplicating more things around

File: devsetup/tripleo/tripleo_install.sh
Comment (bogdando) [reply devsetup/tripleo/tripleo_install.sh:33]: this made it simpler to iterate for multi-stack, where is stack 0 is overcloud.
I will retain the name via adding an alias to the old names for a single-stack multi-node compat

File: devsetup/scripts/tripleo.sh
Comment (marios) [reply devsetup/scripts/tripleo.sh:71]: even with +x i don't think this is acceptable since you are actually writing the user/pass to file here

this script is being executed by the existing multinode ci jobs i.e. not just for your new multi cell addition.

Isn't the whole point of this work that we will be adding a new multi cell job in ci? So what do you mean about local setups.

File: devsetup/scripts/tripleo.sh
Comment (marios) [inline devsetup/scripts/tripleo.sh:125]: same concern here about leaking the registry user/pass

**Review by marios — DISMISSED** [review]

File: devsetup/tripleo/net_config.j2
Comment (marios) [reply devsetup/tripleo/net_config.j2:32]: possibly but lets please wait until that is merged we are very close now and startign to merge the series

we can discuss next week?

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:125]: > even with +x i don't think this is acceptable since you are actually writing the user/pass to file here
> this script is being executed by the existing multinode ci jobs i.e. not just for your new multi cell addition.
> Isn't the whole point of this work that we will be adding a new multi cell job in ci? So what do you mean about  local setups.

As a developer, I need a way to deploy it locally, using this same code.
It won't fly without registry creds. There is no reproducer for it in CI, or any docs on how to use it without auto gold nodes in zuul (not the best choice neither)

Comment (bogdando) [reply devsetup/scripts/tripleo.sh:125]: We can make a patch to ci_framework to always unset ``RH_REGISTRY_PWD`` as adependency of this change. Then it should address your concen

Comment (bogdando) [reply devsetup/scripts/tripleo.sh:125]: Done! Does it make sense now https://review.rdoproject.org/r/c/rdo-jobs/+/53192/34..35/playbooks/data_plane_adoption/install_yamls_deploy_tripleo.yaml ?

File: devsetup/tripleo/net_config.j2
Comment (bogdando) [reply devsetup/tripleo/net_config.j2]: I failled back to a flat subnet, there is yet routing issues for /16 external network to address for that switch in ci framework

**Comment by bogdando** [comment]
Local source trilpeo deployment has completed w/o errors for this revision.

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:127]: PTAL @marios

Comment (marios) [reply devsetup/scripts/tripleo.sh:127]: ack thanks @bogdando. I think https://review.rdoproject.org/r/c/rdo-jobs/+/53192/34..35/playbooks/data_plane_adoption/install_yamls_deploy_tripleo.yaml should make sure we aren't setting this in the ci runs.

It will be confusing though because someone will rightly ask "why are you setting the username and password of the registry to '' " ?

Thanks for acknowledging the concern and sanity checking that we aren't going to run this in ci.

**Review by marios — COMMENTED** [review]
can you please ensure this is tested with the periodic multinode jobs before we merge

i.e. in your testproject run (at least one of)
```
periodic-data-plane-adoption-multinode-osp-17-to-extracted-crc-no-ceph
periodic-data-plane-adoption-multinode-osp-17-to-extracted-crc-with-ceph
```

**Comment by bogdando** [comment]
> can you please ensure this is tested with the periodic multinode jobs before we merge
> 
> i.e. in your testproject run (at least one of)
> 
> ```
> periodic-data-plane-adoption-multinode-osp-17-to-extracted-crc-no-ceph
> periodic-data-plane-adoption-multinode-osp-17-to-extracted-crc-with-ceph
> ```

https://review.rdoproject.org/r/c/testproject/+/53212/8..9/.zuul.yaml

File: devsetup/scripts/tripleo.sh
Comment (marios) [inline devsetup/scripts/tripleo.sh:48]: do we really need this comment, there is an existing variable and indeed what you are documenting here is how it is used

**Review by marios — COMMENTED** [review]
just another pass but its a huge diff 
main concern is the existing multinode jobs but i believe you are already running those in your test runs.

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:48]: I have copied this note from standalone tripleo shell script, to align them better. It makes sense to clarify that important moment here as well

**Comment by bogdando** [comment]
> just another pass but its a huge diff main concern is the existing multinode jobs but i believe you are already running those in your test runs.

there is no huge diff if you follow the "Compare" button updates to show only changes for a revision. Only reabasing brings in irrelevant changes.

**Comment by bogdando** [comment]
This is already deployable locally for a few weeks as of now, yet I am still struggling to make networking configs working in testproject. This is still WIP, just for your info.

**Comment by bogdando** [comment]
Without tripleo-networking applied, I am still having a pingtest problem with external network GW in testproject.
Adding an option to omit ping tests completely (for the srouce tripleo cloud), when going with a BYI network on deployed servers

File: devsetup/tripleo/net_config.j2
Comment (cescgina) [inline devsetup/tripleo/net_config.j2:17]: @bogdando this bit is what is causing the undercloud node to become unreachable

Comment (cescgina) [reply devsetup/tripleo/net_config.j2:17]: this causes a restart of the `eth0` interface, that has the public ip that zuul assigns to the node

**Comment by bogdando** [comment]
New test project for multinode jobs (not deploying multicell yet) https://review.rdoproject.org/r/c/testproject/+/54199

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:50]: I understand this is already published elsewhere, but I don't think we should be posting internal urls in public repos

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:54]: just want to make sure I'm parsing the if correctly, this is trying to check that the number of compute cells is 1 or 3, right?

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:60]: wouldn't it be better to stop the execution at this point if this configuration is unsupported atm?

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:100]: `False` should be lowercase, as that is what we're checking later on https://github.com/openstack-k8s-operators/install_yamls/pull/826/files#diff-6d137e1f40b93f9957103127166db87dbd340c674dcf1d8694def950ebd56821R101

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:101]: same as above

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:124]: should use `$HOME` instead of `/home/zuul`, as this block is mostly intended for dev use cases, where user will likely be different

File: devsetup/scripts/tripleo.sh
Comment (cescgina) [inline devsetup/scripts/tripleo.sh:177]: we should not hardcode the user here

File: devsetup/tripleo/net_config.j2
Comment (cescgina) [inline devsetup/tripleo/net_config.j2:17]: we should include a default false here in case this is not defined (we use this template to configure networking in ci without calling the tripleo.sh script)

**Comment by bogdando** [comment]
Needs https://review.rdoproject.org/r/c/rdo-jobs/+/54221

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:54]: correct

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:119]: perhaps, however there are lots of assumptions have already been made for using zuul user in this and other scripts

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:172]: the issue on my setup was that on hypervisor I had no zuul user, but in VMs, it needed to be zuul. So that default HOME env var was incorrectly interpolated by j2 render.

To properly fix this we'd need to introduce a new env var for these scripts and Make file, somthing like CLOUD_USER, and get rid of all assumptions made here for zuul user, which goes a bit beyond of this patch scope

**Comment by jistr** [comment]
I think we should probably discuss this and https://github.com/openstack-k8s-operators/data-plane-adoption/pull/517 on one of the adoption calls or better yet, schedule a separate call to do a high level walk through and what should be the plan for going forward with multi-cell?

File: devsetup/tripleo/net_config.j2
Comment (marios) [reply devsetup/tripleo/net_config.j2:17]: i have a similar concern about the default_route
I arrived here following the TRIPLEO_NETWORKING
The intention is that this will only affect the multicell scenario/job but the default given is TRUE 

```
EDPM_CONFIGURE_NETWORKING ?= true
```

We could either make the default be false, and as @cescgina pointed out default it here, or the inverse, which is what you proposed with https://review.rdoproject.org/r/c/rdo-jobs/+/54221 where you explicitly set it false for the vanila multinode case

```
            EDPM_CONFIGURE_NETWORKING: false  

```

In my opinion it is more appropriate for this to default to false and then you can explicitly set this in the multi-cell job where is will be needed.

In any case though and before we merge we must run the multinode jobs so we can sanity check they are ok with these changes. I know you've run them in the past but we should run them again once this is ready for merge  as there are many ongoing changes here.

File: devsetup/scripts/standalone.sh
Comment (marios) [inline devsetup/scripts/standalone.sh:125]: fair enough but this change has enough going on without also dealing with this default switch here

File: devsetup/scripts/tripleo.sh
Comment (marios) [inline devsetup/scripts/tripleo.sh:51]: can we instead explicitly search for what IS allowed ? i.e. the "main cell0 plus additional 2 compute cells"

**Review by marios — COMMENTED** [review]
@bogdando I think it may help to coordinate everyone with a call to discuss this and https://review.rdoproject.org/r/c/rdo-jobs/+/53192 and https://github.com/openstack-k8s-operators/ci-framework/pull/1984 and the others in the series

the ci reviews call may be one place, or if you prefer to setup another call?
maybe we can sync about that in today's data plane adoption general/weekly.

Before we start merging and after any further comments are addressed/changes made here or in the rest of the patches we need to make sure we run the multinode jobs (otherwise we'll just be chasing that to fix things)

in upstream we need to run at least one of the jobs (ceph/no ceph)


[periodic-adoption-standalone-to-crc-no-ceph](https://review.rdoproject.org/zuul/job/periodic-adoption-standalone-to-crc-no-ceph)
[periodic-adoption-standalone-to-crc-ceph](https://review.rdoproject.org/zuul/job/periodic-adoption-standalone-to-crc-ceph)

downstream please run the networker job 
[periodic-internal-adoption-multinode-networker-to-crc](https://sf.hosted.upshift.rdu2.redhat.com/zuul/t/tripleo-ci-internal/job/periodic-internal-adoption-multinode-networker-to-crc)

**Comment by bogdando** [comment]
the [failure](https://logserver.rdoproject.org/99/54199/4/check/periodic-adoption-multinode-to-crc-no-ceph/b2673be/controller/ci-framework-data/logs/ci_script_010_run_tripleo.log) in testproject for existing multinode jobs looks real
``" Cannot find any hosts on "overcloud" in network "ctlplane": tripleoclient.exceptions.DeploymentError: Cannot find any hosts on "overcloud" in network "ctlplane"``

File: devsetup/scripts/standalone.sh
Comment (bogdando) [reply devsetup/scripts/standalone.sh:125]: could you elaborate? in rod jobs we provide an overriding interface https://github.com/29418/rdo-jobs/blob/6c8849c108a11ff1167ad9e726aeb5f039b9edfb/playbooks/data_plane_adoption/ci_framework_deploy_standalone_vm.yaml#L12 so I don't get the question

**Comment by bogdando** [comment]
> @bogdando I think it may help to coordinate everyone with a call to discuss this and https://review.rdoproject.org/r/c/rdo-jobs/+/53192 and [openstack-k8s-operators/ci-framework#1984](https://github.com/openstack-k8s-operators/ci-framework/pull/1984) and the others in the series

for simplicity, I would defer discussing next steps, which is https://review.rdoproject.org/r/c/rdo-jobs/+/53192 and https://github.com/openstack-k8s-operators/ci-framework/pull/1984, and https://github.com/openstack-k8s-operators/data-plane-adoption/pull/517 . This patch scope if a fasts and quick way to stand up a development setup of multi-stack tripleo. This is needed for futher work until we get to a workable ci-framework "osp-plugin", with multi-stack, solution

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:51]: do you mean state in the echo message? what are we going to "search" here?

**Comment by bogdando** [comment]
> Before we start merging and after any further comments are addressed/changes made here or in the rest of the patches we need to make sure we run the multinode jobs

it should be good as well to make sure it passes test project https://review.rdoproject.org/r/c/testproject/+/54199

File: devsetup/scripts/standalone.sh
Comment (marios) [reply devsetup/scripts/standalone.sh:125]: you are changing the default. The change itself is probably right since we generally avoid exposing the interal urls in this space. It does mean though that all jobs previously overriding the default will now have to switch it. I am questioning if that is a necessary change as part of the series 
i.e. it is 'right' but is it needed here, on top of all the other changes being made.

File: devsetup/scripts/tripleo.sh
Comment (marios) [inline devsetup/scripts/tripleo.sh:84]: why don't we just add this file into the rest of the env files instead of creating like that
i mean, what is special about this one?

File: devsetup/scripts/tripleo.sh
Comment (marios) [inline devsetup/scripts/tripleo.sh:109]: what about the case where we have the user/pwd set to ""  i.e. like in https://review.rdoproject.org/r/c/rdo-jobs/+/54221/5/playbooks/data_plane_adoption/install_yamls_deploy_tripleo.yaml
will it still work if you are using --enable-registry-login?

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:84]: it is a recommended approach in general to provide separate files for different configuration areas, rather than mangling large files.

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:109]: I guess https://review.rdoproject.org/r/c/testproject/+/54199 will show this out

File: devsetup/scripts/tripleo.sh
Comment (marios) [reply devsetup/scripts/tripleo.sh:51]: like if EDPM_COMPUTE_CELLS != 3 (or whatever the allowed is i think 3?) then error

**Review by marios — COMMENTED** [review]
waiting to hear that testing is good and will revisit

File: devsetup/scripts/tripleo.sh
Comment (bogdando) [reply devsetup/scripts/tripleo.sh:51]: This already explicitely checks for allowed values of a 1 and 3 :)

**Comment by bogdando** [comment]
the regression test is passed the multi-node check https://review.rdoproject.org/r/c/testproject/+/54199 ! Now please help testing it downstream, for networker

**Comment by bogdando** [comment]
recheck

File: devsetup/scripts/standalone.sh
Comment (marios) [reply devsetup/scripts/standalone.sh:125]: i just sanity checked that we are already setting this to d/stream servers for our dstream jobs so we should be OK on the default switch.
marking resolved.

File: devsetup/scripts/tripleo.sh
Comment (marios) [reply devsetup/scripts/tripleo.sh:84]: i meant more like, we are already carrying a bunch of environment files. Fine if you don't want to put this into an existing one but you can just add a new file under the tripleo directory. 
With this we have some environments/parameter_defaults coming from files, and others like this one being created by heredoc. I think it just further adds to the complexity of this 'tool'.

File: devsetup/scripts/tripleo.sh
Comment (marios) [reply devsetup/scripts/tripleo.sh:109]: result looks OK 

https://logserver.rdoproject.org/99/54199/4/check/periodic-adoption-multinode-to-crc-no-ceph/8770982/undercloud/zuul/containers-prepare-parameters.yaml

i guess the next bit for adding creds into the file is not executed since we set user/pass to ""

**Review by marios — APPROVED** [review]
networker log looks OK with tripleo deploy complete https://sf.hosted.upshift.rdu2.redhat.com/logs/78/678/02dce54df035fc84769dd16020946b926ebd7a51/check-gitlab-cee/periodic-internal-adoption-multinode-networker-to-crc/26a0f31/controller/ci-framework-data/logs/ci_script_010_run_tripleo.log

Overcloud Deployed successfully

**Comment by bogdando** [comment]
can we merge this now please?

**Comment by bogdando** [comment]
@jistr PTAL

File: devsetup/scripts/common.sh
Comment (olliewalsh) [inline devsetup/scripts/common.sh:45]: I would avoid using different logic to the ansible bool filter. It will return True for "yes", False for "no", while `bool("no")` will return True. https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/filter/core.py#L89

File: devsetup/scripts/tripleo.sh
Comment (olliewalsh) [reply devsetup/scripts/tripleo.sh:51]: `if [ $EDPM_COMPUTE_CELLS -ne 1 -a $EDPM_COMPUTE_CELLS -ne 3 ]; then ...`

File: devsetup/scripts/tripleo.sh
Comment (olliewalsh) [inline devsetup/scripts/tripleo.sh:134]: where is this used?

File: devsetup/scripts/common.sh
Comment (olliewalsh) [reply devsetup/scripts/common.sh:45]: Also expect filtering through bool will be a no-op. `if <expr>` and `if bool(<expr>)` are equivalent AFAIK

**Comment by fao89** [comment]
/hold
just noticed @olliewalsh comments

**Comment by bogdando** [comment]
@olliewalsh I will address your comments in follow up, sorry.

File: devsetup/scripts/common.sh
Comment (bogdando) [reply devsetup/scripts/common.sh:45]: This is not a perfect solution, I agree. However, I tested it locally and in CI for multinode jobs, and it works. I will try to improve this in follow ups

File: devsetup/tripleo/tripleo_install.sh
Comment (marios) [inline devsetup/tripleo/tripleo_install.sh:17]: @bogdando please check slack when you get some time

this breaks the ceph job (your test ran multinode-no-ceph) so please have a look and if we cannot find a good solution today then we can revert and work it out before going again

**Pull request merged**
